tuf/encoding/metadata_definitions.asn1

```diff
@@ -0,0 +1,302 @@
+/*
+This file defines the ASN.1 format for TUF metadata in a language-neutral,
+platform-neutral manner.
+
+This is not the *internal* format used in this TUF implementation, but it is
+supported. Please see the TUF specification for the JSON format and more
+detailed explanation of TUF metadata.
+
+The TUF implementation in this repository uses modules in tuf.encoding and
+the Python library pyasn1 to construct, convert, encode, and decode metadata
+using ASN.1 and ASN.1/DER.
+
+
+NO: CURRENTLY SKIPPING THIS STEP:
+  We compile this abstract file into a slightly lower level version of this
+  file (with IMPLICIT markers and sequence positions) using asn1c and then
+
+We convert this file into a Python module, metadata_asn1_definitions.py, which
+pyasn1 and those modules can use. The instructions for doing this are included
+at the top of metadata_asn1_definitions.py
+
+
+This file is organized into five sections:
+   Common Definitions
+   Root Role Definitions
+   Timestamp Role Definitions
+   Snapshot Role Definitions
+   Targets Role Definitions
+
+
+A note for implementers of TUF-specification-conformant systems:
+There are constraints and optimizations that you can bake into your ASN.1
+definitions. They are not included here for readability and simplicity.
+
+Here are some example constraints and optimizations:
+  Length Constraints
+ Filename        ::= VisibleString (SIZE(1..256))
+
+  Character Type constraint example: disallow '/' and '\'.
+ StrictFilename  ::= VisibleString (SIZE(1..32))] (PATTERN "[^/\\]+")
+ URLs            ::= SEQUENCE (SIZE(0..256)) OF URL
+ URL             ::= VisibleString (SIZE(1..1024))
+
+  Enumerations (optimization)
+    You could swap in enums in place of some of the values that currently
+    use strings. For example:
+  RoleType        ::= ENUMERATED {root, targets, snapshot, timestamp}
+  HashFunction ::= ENUMERATED {sha224, sha256, sha384, sha512, sha512-224,
+                               sha512-256, ...}
+  SignatureMethod ::= ENUMERATED {rsassa-pss, ed25519, ...}
+
+
+A note for any future TUF/upTUF/Uptane repository maintainers:
+ I'm excluding character type constraints at this stage. These constraints
+ were always excluded in the  final .py module in any case (presumably due
+ to compilation issues), so the only change is the removal of
+ StrictFilename, which was functionally equivalent to Filename in any event.
+ I'm also removing length constraints and enumerations to keep this simple
+ for now. Implementers will impose their own length constraints in their
+ ASN.1 definitions (or, if they are able, at a higher level) and perform
+ their own optimizations.
+
+
+Questions:
+ - Would it be better to use fewer empty wrapped classes? e.g. is it better to
```

Comment:
I'm personally not a fan of abstraction which isn't used for some form of data hiding. My vote would be towards yes, less empty classes.